include/storage/flash_map.h

```diff
@@ -113,14 +115,17 @@ int flash_area_check_int_sha256(const struct flash_area *fa,
  *
  * Function Retrieves flash_area from flash_map for given partition.
  *
- * @param[in]  id ID of the flash partition.
+ * @param[in]  id ID of the flash partition. The id is checked if it is within
+ *             range of __flash_map linker section. If yes, it is considered
+ *	       to be directo pointer to flash_area definition and returned;
+ *	       otherwise the __flash_map will be searched for the ID.
  * @param[out] fa Pointer which has to reference flash_area. If
  * @p ID is unknown, it will be NULL on output.
  *
  * @return  0 on success, -EACCES if the flash_map is not available ,
  * -ENOENT if @p ID is unknown.
  */
-int flash_area_open(uint8_t id, const struct flash_area **fa);
+int flash_area_open(const uintptr_t id, const struct flash_area **fa);
```

Comment:
Changing flash_area_open() to accept a pointer feels a little weird and potentially a bit fragile.

It seems like we could add a separate function, like this:

```c
   /* maybe a name could be found than "..._ptr" */
   int flash_area_open_ptr(struct flash_area *area);
```

This would return 0 if `area` was successfully opened, and `-errno` otherwise.

Why do we have to make this change? The commit log says what is happening, but doesn't explain why.

Same question for `flash_area_get_sectors()`.

I'm guessing this is somehow for compatibility with mcuboot, but I thought one goal of this PR was to add some macros in mcuboot which wrap the Zephyr APIs with what MCUboot needs, so I don't get why this is needed in Zephyr; why is it?

Reply:
Another related question is why we are keeping around flash_area_open(), since we don't actually do anything on close(), instead of just replacing it with a  `flash_area_from_id` function like I ask about below. We could keep flash_area_open() around for compatibility, but deprecate it.